functionalDB/chapter.md

```diff
@@ -0,0 +1,1138 @@
+         [k1 l2map] index  ; keys and values of the first level
+         :when (try (lvl1-prd k1) (catch Exception e false))  
+         [k2  l3-set] l2map  ; keys and values of the second level
+         :when (try (lvl2-prd k2) (catch Exception e false))
+         :let [res (set (filter lvl3-prd l3-set))] ]
+     (with-meta [k1 k2 res] (meta pred-clause))))
+````
+Assuming the query was executed on July 4th, the results of executing it on the above data are:
+<table>
+<tr>
+<td>Result Clause</td><td>Result Meta</td>
+</tr>
+<tr>
+<td>[:likes Pizza #{1}]</td><td>["?e" nil nil]</td>
+</tr>
+<tr>
+<td>[:name USA #{1}]</td><td>["?e" nil "?nm"]</td>
+</tr>
+<tr>
+<td>[:speak "English" #{1, 3}]</td><td>["?e" nil nil]</td>
+</tr>
+<tr>
+<td>[:birthday "July 4, 1776" #{1}]</td><td>["?e" nil "?bd"]</td>
+</tr>
+<tr>
+<td>[:name France #{2}]</td><td>["?e" nil "?nm"]</td>
+</tr>
+<tr>
+<td>[:birthday "July 14, 1789" #{2}]</td><td>["?e" nil "?bd"]</td>
+</tr>
+<tr>
+<td>[:name Canada #{3}]</td><td>["?e" nil "?nm"]</td>
+</tr>
+<tr>
+<td>[:birthday "July 1, 1867" {3}]</td><td>["?e" nil "?bd"]</td>
+</tr>
+</table>
+Once we have produced all of the result clauses, we need to perform an *AND* operation between them. This is done by finding all of the elements that passed all the predicate clauses:
+
+````clojure
+(defn items-that-answer-all-conditions [items-seq num-of-conditions]
+   (->> items-seq ; take the items-seq
+         (map vec) ; make each collection (actually a set) into a vector
+         (reduce into []) ;reduce all the vectors into one vector
+         (frequencies) ; count for each item in how many collections (sets) it was in
+         (filter #(<= num-of-conditions (last %))) ;items that answered all conditions
+         (map first) ; take from the duos the items themselves
+         (set))) ; return it as set
+````
+
+In our example, the result of this step is a set that holds the value *1* (which is the entity Id of USA). 
+
+We now have to remove the items that didn’t pass all of the conditions:
+
+````clojure
+(defn mask-path-leaf-with-items [relevant-items path]
+     (update-in path [2] CS/intersection relevant-items))
+````
+
+Finally, we remove all of the result clauses that are 'empty' (i.e. their last item is empty.) We do this in the last line of the *query-index* function. Our example leaves us with the following items:
+
+<table>
+<tr>
+<td>Result Clause</td><td>Result Meta</td>
+</tr>
+<tr>
+<td>[:likes Pizza #{1}]</td><td>["?e" nil nil]</td>
+</tr>
+<tr>
+<td>[:name USA #{1}]</td><td>["?e" nil "?nm"]</td>
+</tr>
+<tr>
+<td>[:birthday "July 4, 1776" #{1}]</td><td>["?e" nil "?bd"]</td>
+</tr>
+<tr>
+<td>[:speak "English" #{1}]</td><td>["?e" nil nil]</td>
+</tr>
+</table>
+We are now ready to being reporting the results. The result clause structure is unwieldy for this purposes, so we will convert it into an an index-like structure (map of maps) -- with a significant twist. 

```

Comment:
unwieldy for this purposes -> unwieldy for *these* purposes **OR** unwieldy for this *purpose*